src/maelstrom/workload/lock.clj

```diff
@@ -0,0 +1,79 @@
+(ns maelstrom.workload.lock
+  "A distributed lock workload: sends lock and unlock messages."
+  (:require [maelstrom [client :as c]]
+            [jepsen [checker :as checker]
+                    [client :as client]
+                    [generator :as gen]]
+            [schema.core :as s]))
+
+(c/defrpc lock!
+  "Clients send `lock` messages to servers. Servers should respond with
+  `lock_ok` messages when the lock is granted."
+  {:type (s/eq "lock")}
+  {:type (s/eq "lock_ok")})
+
+(c/defrpc unlock!
+  "Clients send `unlock` messages to servers. Servers should respond with
+  `unlock_ok` messages acking that the lock is released"
+  {:type (s/eq "unlock")}
+  {:type (s/eq "unlock_ok")})
+
+(defn client
+  ([net]
+   (client net nil nil))
+  ([net conn node]
+   (reify client/Client
+     (open! [this test node]
+       (client net (c/open! net) node))
+
+     (setup! [this test])
+
+     (invoke! [_ test op]
+       (c/with-errors op #{}
+         (let [rpc (case (:f op) :lock lock! :unlock unlock!)
+               timeout (* 1000 (:time-limit test))] ; timeout at most last op from each client
```

Comment:
This is a sort of unusual choice--usually you'd want to choose something relatively small here. Could you talk more about this?

Reply:
Contrary to most (all?) workloads, this is for a blocking abstraction, where the reply to an operation depends on requests from other clients. I.e., the lock operation should only be replied to when the lock is granted. So, this workload chooses rpc timeout such that the test will never timeout, unless near the end, when no more ops are being issued. This is to achieve simplicity; otherwise the server(s) could think a client got the lock when the client has the op failed. This is for an exercise where failures are not the focus.

Reply:
Ohhhhhh you don't reply with an error on lock acquisition failure, you actually hold the client's RPC open indefinitely. Huh. I don't think I want to do this in maelstrom's official repo--it makes it easy for students to write incorrect lock services and for the test to pass without finding safety violations.

Reply:
There is no locking acquisition failure. That is why the client is not supposed to timeout (and the choice for timeout in the workload). This workload is not supposed to be for real services, but to play with / understand algorithms. That is why it does not inject failures. :) Yes, we have been using Maelstrom for this blocking abstraction, illustrating the run of an algorithm like the distributed lock in the classic Lamport 1978 paper. I am not saying that a distributed lock is a good idea, but should Maelstrom be limited to teaching non-blocking abstractions, or should we be able to use it for scenarios where a request is fulfilled  when some other client invokes some other op (like the classic bounded-buffer), without having to do busy-waiting repeating RPCs until success? I guess this is the central issue. Think about it. If you are not comfortable with such things in the repo, then there is no point in pursuing this proposal. (But it would be a pity. Is has been very pleasant watching Lamport diagrams of Lamport's distributed lock.)

Reply:
It's cool that this works as a demo of your particular lock service, but the point of Maelstrom is to test that students implementations work correctly, and this tactic isn't going to be as good at that. It's not that you *can't* write indefinitely-blocking tests; it's that blocking tests tend to mask errors that nonblocking tests would be able to detect. That's why essentially every Jepsen test and Maelstrom workload use bounded-blocking (e.g. a few seconds tops) or nonblocking operations.

Reply:
It is a particular, but important case. I agree that the idea is to test that implementations work correctly ... for a given system model.  What was missing, describing the workload, is that this is to test distributed lock algorithms for "asynchronous systems with no failures". Before students start thinking about failures and partitions, they may play with algorithms that will simply not work if there are failures, such as the one mentioned above from Lamport, or the Ricart-Agrawala one. Contrary to Jepsen, where we want to test real systems, here we may use Maelstrom not thinking about a real system, but e.g., to check and animate an N-node algorithm, where "clients" are not even real "remote clients", but a means to an end (inject load in the node and probe what is going on) representing a local thread in the node. For this "distributed lock assuming no failures" scenario, "infinite" timeouts, allowing the use of the natural blocking abstraction, with a load injector alternating between lock-unlock is the best way for students to learn such algorithms without getting distracted with non-essential details. (Later when moving to failures, other examples than locking are used.)

Now the issue is, could the above workload be made available, and properly documented, as one of possible workloads, this one devoted to a system model with no failures, or there should be a single lock workload? Assuming that there would be a single lock workload in Maelstrom, then it should not be restricted to that "no failures" system model, but cover the general case. I can think about it and how the checker would be (probably using the checker/linearizable, as you suggest), but I don't know when I can afford to do it; also I will have to learn more about Maelstrom.

But what I fear is that it will become much more complex, not just the checking but the its spec, and the description of what nodes are supposed to do. E.g., because lock requests can timeout, and be reissued, a server must reply "lock_ok" a second time to that client (not because it is a reentrant lock, which it isn't); if an unlock request can timeout (due to latency or message loss) and be re-issued, the server will have to reply "unlock_ok" (or some variation) a second time, even if by then the lock has been acquired by someone else, which means that the server cannot report an error "of release without having lock".  So, describing the requirements will become ugly, all simplicity lost. To test the general case, the load injection will also have to be more general, which together with timeouts will cause unwieldy space-time diagrams, making learning more difficult, deviating from the essentials.

So, there is a tradeoff, wether the goals are testing real lock services (which would definitely be Jepsen goal), or providing a workload for learning classic toy distributed lock algorithms for a system model with no failures, which was my intention with this workload.